### PR TITLE
Add support for dropdown answer in detail answer

### PR DESCRIPTION
**Pull request by MebinAbraham**
### What is the context of this PR?
- Adds support for dropdown answer types in detail answer.
- Adds support to override default select field placeholder text.
- Initial renames of detail answer schemas to improve consistency.

### How to review 
Test using `test_checkbox_detail_answer_dropdown.json` or `test_radio_detail_answer_dropdown.json`

- Ensure a dropdown in a detail answer behaves as expected.

Requires:
- https://github.com/ONSdigital/eq-questionnaire-validator/pull/82
- https://github.com/ONSdigital/eq-translations/pull/71

### Checklist

* [ ] New static content marked up for translation
* [x] Newly defined schema content included in eq-translations repo

### CODE REVIEW

File: app/jinja_filters.py [hunk 1]
```diff
@@ -357,15 +369,14 @@ def map_relationships_config_processor():
 
 class SelectOptionConfig:
     def __init__(self, option, select):
-        self.text = option[1]
-        self.value = option[0]
+        self.value, self.text = option
```
Comment (ajmaddaford) [inline app/jinja_filters.py:372]: 👌

File: tests/app/test_jinja_filters.py [hunk 2]
```diff
@@ -202,12 +202,17 @@ def test_radio_class_visible_attribute():
...
+                        "id": "mandatory-checkbox-with-mandatory-dropdown-detail-answer",
+                        "mandatory": True,
                         "visible": True,
+                        "label": "Please specify heat level",
+                        "placeholder": "Select heat level",
+                        "options": [
+                            {"label": "Mild", "value": "Mild"},
+                            {"label": "Medium", "value": "Medium"},
+                            {"label": "Hot", "value": "Hot"},
+                        ],
```
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py:215]: I can't see why this test was changed to a dropdown when it's the visible property that is being tested?

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    it("When the user selects an option with an optional detail answer but does not provide a detail answer, Then the summary should display the chosen option without the detail answer", () => {
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.submit()).click();
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Fruit");
+    });
+
+    it("When the user selects an option with an optional detail answer and provides a detail answer, Then the summary should display the chosen option and its details", () => {
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.fruitDetail()).selectByAttribute("value", "Mango");
+      expect($(CheckboxDropdownPage.fruitDetail()).getValue()).to.equal("Mango");
```
Comment (ajmaddaford) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:35]: I think this `expect` is unnecessary.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      $(CheckboxDropdownPage.submit()).click();
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Fruit\nMango");
+    });
+
+    it("When the user selects the default dropdown option after submitting a detail answer, Then the summary should not display the detail answer", () => {
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.fruitDetail()).selectByAttribute("value", "Mango");
+      $(CheckboxDropdownPage.submit()).click();
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Fruit\nMango");
```
Comment (ajmaddaford) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:44]: I think this `expect` is unnecessary.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    it("When the user selects an option with an mandatory detail answer but does not provide a detail answer, Then an error should be displayed when the user submits", () => {
+      $(CheckboxDropdownPage.jam()).click();
+      $(CheckboxDropdownPage.submit()).click();
+      expect($(DropdownMandatoryPage.errorNumber(1)).getText()).to.equal("Please select the type of Jam");
+    });
+
+    it("When the user selects an option with an mandatory detail answer and provides a detail answer, Then the summary should display the chosen option and its details", () => {
+      $(CheckboxDropdownPage.jam()).click();
+      $(CheckboxDropdownPage.jamDetail()).selectByAttribute("value", "Strawberry");
+      expect($(CheckboxDropdownPage.jamDetail()).getValue()).to.equal("Strawberry");
```
Comment (ajmaddaford) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:61]: I think this `expect` is unnecessary.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      $(CheckboxDropdownPage.submit()).click();
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Jam\nStrawberry");
+    });
+
+    it("When the user removes a previously submitted detail answer, Then the summary should not display the removed detail answer", () => {
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.fruitDetail()).selectByAttribute("value", "Mango");
+      $(CheckboxDropdownPage.submit()).click();
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Fruit\nMango");
```
Comment (ajmaddaford) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:70]: I think this `expect` is unnecessary (this was tested in an earlier test).

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+      $(SummaryPage.previous()).click();
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.submit()).click();
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("No answer provided");
+    });
+
+    it("When the user selects multiple options with detail answers and submits, Then the summary should display the sll the chosen options and their details", () => {
```
Comment (ajmaddaford) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:78]:
```suggestion
    it("When the user selects multiple options with detail answers and submits, Then the summary should display all the chosen options and their details", () => {
```

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.fruitDetail()).selectByAttribute("value", "Mango");
+      $(CheckboxDropdownPage.jam()).click();
+      $(CheckboxDropdownPage.jamDetail()).selectByAttribute("value", "Strawberry");
+
+      expect($(CheckboxDropdownPage.jamDetail()).getValue()).to.equal("Strawberry");
+      expect($(CheckboxDropdownPage.jamDetail()).getValue()).to.equal("Strawberry");
```
Comment (ajmaddaford) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:85]: I think these `expect` statements are unnecessary.

File: tests/functional/spec/components/radio/radio_detail_answer_dropdown.spec.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    it("When the user selects an option with an optional detail answer but does not provide a detail answer, Then the summary should display the chosen option without the detail answer", () => {
+      $(RadioDropdownPage.fruit()).click();
+      $(RadioDropdownPage.submit()).click();
+      expect($(SummaryPage.optionalRadioWithDropdownDetailAnswer()).getText()).to.equal("Fruit");
+    });
+
+    it("When the user selects an option with an optional detail answer and provides a detail answer, Then the summary should display the chosen option and its details", () => {
+      $(RadioDropdownPage.fruit()).click();
+      $(RadioDropdownPage.fruitDetail()).selectByAttribute("value", "Mango");
+      expect($(RadioDropdownPage.fruitDetail()).getValue()).to.equal("Mango");
```
Comment (ajmaddaford) [inline tests/functional/spec/components/radio/radio_detail_answer_dropdown.spec.js:35]: The same comments about unnecessary expects in the middle of a test apply to the tests in this file as well.

File: tests/app/test_jinja_filters.py (hunk 2 above)
Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py:215]: It was simply to satisfy 1 line of coverage to avoid writing a suite of integration tests when there are functional tests for it. Didn't think to add more jinja filter tests since none currently exist to test the config anything but like an individual attribute. I could technically add new tests to test the full context.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js (hunk 3 above)
Comment (MebinAbraham) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:35]: It was just to verify the dropdown itself worked and not shown on summary due to some other reason but can be removed if the answer on summary is enough to indicate it works.

Comment (ajmaddaford) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:35]: I think the test is testing the summary, and extra `expect` statements cloud the intent of the test.

File: tests/app/test_jinja_filters.py (hunk 2 above)
Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py:215]: Just cover the ```if answer_type == "Dropdown":` block.

Comment (ajmaddaford) [reply tests/app/test_jinja_filters.py:215]: Can't we just have an additional specific test for other dropdown answers?

Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py:215]: Ya my point was we dont have specific tests for testfield/numeric etc, only testing the "open" attribute really. Let me refactor the existing ones to reuse the same partial json and write some to test each variant.

Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py]: Added some tests to test the full `OtherConfig` independently, let me know what you think.

File: tests/app/test_jinja_filters.py [hunk 10]
```diff
@@ -460,3 +375,106 @@ def test_format_address_fields_with_uprn():
...
+        questionnaire_schema,
+        question_schema,
+        answer_store=Mock(),
+        metadata={},
+        location=None,
+        data=None,
+        form_data=None,
+    )
+
+    return getattr(form, answer_schema["id"])
```
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py:394]: We shouldn't be introducing dependencies on some quite involved classes. In the previous tests the form was mocked.

File: tests/app/test_jinja_filters.py [hunk 11]
```diff
@@ -460,3 +375,106 @@ def test_format_address_fields_with_uprn():
...
+    [
+        (True, True),
+        (False, False),
+        (None, False),
+    ],
+)
+def test_dropdown_detail_answer_config(
+    app, answer_schema_dropdown, is_visible, expected_visibility
+):
+    _test_detail_answer_config(answer_schema_dropdown, is_visible, expected_visibility)
```
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py:480]: I find these tests hard to follow. They are parametrized to test for visibility, but test all properties of `OtherConfig`. 

I think the tests for visibility should be separated from the tests for the detail answer types. Also, as the visibility logic is the same for all detail answer types, we should only need to test one type

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

File: tests/app/test_jinja_filters.py (hunk 10 above)
Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py:394]: Ya its because previous the form wasn't being used in the test, i.e only visibility was tested. Maybe I should only test visibility and input type etc giving these are exisiting.

File: tests/app/test_jinja_filters.py (hunk 11 above)
Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py:480]: I'll make 1 test for visibility and 1 tests for the input type check.

Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py]: Have pushed some tests that should be easier to digest.

File: tests/app/test_jinja_filters.py [hunk 12]
```diff
@@ -1,12 +1,15 @@
 # coding: utf-8
 from unittest.mock import MagicMock, patch
 
+import pytest
 from jinja2 import Undefined
 from mock import Mock
 
+from app.forms import error_messages
+from app.forms.field_handlers import get_field_handler
+from app.forms.questionnaire_form import generate_form
```
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py:10]: I think these imports can go now.

File: tests/app/test_jinja_filters.py [hunk 13]
```diff
@@ -455,3 +375,63 @@ def test_format_address_fields_with_uprn():
...
+        (None, 10),
+        (1, 1),
+        (123123123123, 20),
+    ],
+)
+def test_numeric_detail_answer_input_class_config(
+    answer_schema_number, min_or_max_value, expected_width
+):
+    if min_or_max_value:
+        answer_schema_number["maximum"] = {"value": min_or_max_value}
```
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py:392]: If `min_or_max_value` is always `maximum`, then it should be named `max_value`.

File: tests/app/test_jinja_filters.py [hunk 14]
```diff
@@ -455,3 +375,63 @@ def test_format_address_fields_with_uprn():
...
+
+@pytest.mark.parametrize(
+    "min_or_max_value, expected_width",
+    [
+        (None, 10),
+        (1, 1),
+        (123123123123, 20),
+    ],
+)
+def test_numeric_detail_answer_input_class_config(
```
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py:388]: I know there is some inconsistencies with the naming of other/detail answers, but this test is testing the `OtherConfig` class, and so the tests should be named as such.

Also, I think for consistency and so the grouping of tests is clearer, the test names should all start with a consistent name e.g. `test_other_config_`.

File: tests/app/test_jinja_filters.py [hunk 15]
```diff
@@ -455,3 +375,63 @@ def test_format_address_fields_with_uprn():
...
+        (None, False),
+    ],
+)
+def test_visible_detail_answer_config(
+    answer_schema_textfield, is_visible, expected_visibility
+):
+    if is_visible is True:
+        answer_schema_textfield["visible"] = True
+    elif is_visible is False:
+        answer_schema_textfield["visible"] = False
```
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py:434]:
```suggestion
    if is_visible is not None:
        answer_schema_textfield["visible"] = is_visible
```

File: tests/app/test_jinja_filters.py (hunk 13 above)
Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py]: Done

File: tests/app/test_jinja_filters.py (hunk 14 above)
Comment (MebinAbraham) [reply tests/app/test_jinja_filters.py]: Done

**Review by ajmaddaford — APPROVED** [review]

File: scripts/run_validator.sh [hunk 16]
```diff
@@ -1,4 +1,4 @@
 #!/usr/bin/env bash
-tag=latest
+tag=add-support-for-dropdown-placeholder
```
Comment (MebinAbraham) [inline scripts/run_validator.sh:2]: Reminder to revert.
```suggestion
tag=latest
```

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    });
+
+    it("When the user selects an option with an optional detail answer but does not provide a detail answer, Then the summary should display the chosen option without the detail answer", () => {
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.submit()).click();
+
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Fruit");
+    });
+
+    it("When the user selects an option with an optional detail answer and provides a detail answer, Then the summary should display the chosen option and its details", () => {
```
Comment (LJBabbage) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:36]: To be consistent  with your other tests it should be 'detail answer' and not details

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+      $(CheckboxDropdownPage.fruit()).click();
+      $(CheckboxDropdownPage.fruitDetail()).selectByAttribute("value", "Mango");
+      $(CheckboxDropdownPage.submit()).click();
+
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Fruit\nMango");
+    });
+
+    it("When the user selects the default dropdown option after submitting a detail answer, Then the summary should not display the detail answer", () => {
```
Comment (LJBabbage) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:44]: This doesn't read like the others (i.e When the user selects an option with an optional detail answer)

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+      $(CheckboxDropdownPage.fruitDetail()).selectByAttribute("value", "Mango");
+      $(CheckboxDropdownPage.submit()).click();
+      $(SummaryPage.previous()).click();
+      $(CheckboxDropdownPage.fruitDetail()).selectByVisibleText("Select fruit");
+      $(CheckboxDropdownPage.submit()).click();
+
+      expect($(SummaryPage.optionalCheckboxWithDropdownDetailAnswer()).getText()).to.equal("Fruit");
+    });
+
+    it("When the user selects an option with an mandatory detail answer but does not provide a detail answer, Then an error should be displayed when the user submits", () => {
```
Comment (LJBabbage) [inline tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:55]: As above

**Review by LJBabbage — CHANGES_REQUESTED** [review]
It works as expected, but I am struggling with the tests. There seems to be a lot of  functional tests, which some are not needed or could be integration. The ones there could do with a quick second read

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 20]
```diff
@@ -0,0 +1,91 @@
+import CheckboxDropdownPage from "../../../generated_pages/checkbox_detail_answer_dropdown/optional-checkbox-with-dropdown-detail-answer-block.page";
+import SummaryPage from "../../../generated_pages/checkbox_detail_answer_dropdown/summary.page";
+import DropdownMandatoryPage from "../../../generated_pages/dropdown_mandatory/dropdown-mandatory.page";
+
+describe("Optional Checkbox with a Dropdown detail answer", () => {
+  beforeEach(() => {
+    browser.openQuestionnaire("test_checkbox_detail_answer_dropdown.json");
+  });
+
+  describe("Given an optional checkbox with a dropdown detail answer", () => {
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:10]: Fixed the "Then then" typo and used your example.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    it("When a placeholder choice is provided for the detail answer dropdown, Then then provided placeholder should be displayed as the first option'", () => {
+      $(CheckboxDropdownPage.fruit()).click();
+
+      expect($(CheckboxDropdownPage.fruitDetail()).getText()).to.contain("Select fruit");
+    });
+
+    it("When a placeholder choice is not provided for the detail answer dropdown, Then the default placeholder should be displayed as the first option'", () => {
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:17]: Removed the word "dropdown". As for set/provided, I don't think we have a pattern for it, as long as it reads right, I think it's fine.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+      $(CheckboxDropdownPage.jam()).click();
+
+      expect($(CheckboxDropdownPage.jamDetail()).getText()).to.contain("Select an answer");
+    });
+
+    it("When the user does not provide an answer and submits, Then the summary should display 'No answer provided'", () => {
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:23]: Only added to ensure that "detail answer" hasn't caused any issues to the rendering of answers.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js (hunk 17 above)
Comment (MebinAbraham) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:36]: Done

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js (hunk 18 above)
Comment (MebinAbraham) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:44]: Not quite sure what you mean. What do you think it should be?

**Comment by MebinAbraham** [comment]
> It works as expected, but I am struggling with the tests. There seems to be a lot of functional tests, which some are not needed or could be integration. The ones there could do with a quick second read. I stopped short of the 2nd load of functional tests

I avoided integration tests on purpose. We shouldn't write integration test unless it can't be covered by unit tests which this can. The added functional test is because this is a feature and all features should have functional tests.

File: tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js (hunk 18 above)
Comment (LJBabbage) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:44]: When the user selects an option with an optional detail answer and doesn't change the value from the default ...

Comment (MebinAbraham) [reply tests/functional/spec/components/checkbox/checkbox_detail_answer_dropdown.spec.js:44]: The importance here is that they also already previously chose one, so what do you suggest fitting that phrase in?. In summary, the test is, select a non default detail answer, go back and select the default value, and the answer should be removed.

**Review by LJBabbage — APPROVED** [review]
As discussed, this shouldn't have produced 27 functional test. However I will approve.
